[Feature] Added `EuiCollapsibleNav` component

cchaos opened the pull request:
# Feature branch

This is the primary container for the new EuiCollapsibleNav component. It doesn't do much but 
- extend EuiFlyout, 
- push to the left side,
- add `docked` ability, and 
- setup some responsive behavior

## Basic flyout behavior

![Screen Recording 2020-03-03 at 06 40 PM](https://user-images.githubusercontent.com/549577/75830559-2454fa00-5d7f-11ea-86b9-74480cc79f1f.gif)


## Responsive behavior with docking

![Screen Recording 2020-03-03 at 06 40 PM](https://user-images.githubusercontent.com/549577/75830497-0091b400-5d7f-11ea-8d4c-69f0bca32973.gif)

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [x] Props have proper **autodocs**
- [x] Added **documentation** examples
- [x] Added or updated **jest tests**
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- ~[ ] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately~ Will be added via the Feature Branch

cchaos commented:
@myasonik What kind of `aria-alert` or whatever do you think we will need for when the user initiates the opening of the navigation drawer?

@thompsongl Is there any way to add tests for the responsive behavior? Can it mock setting the window to a certain size?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2977/

snide commented on src/components/collapsible_nav/_collapsible_nav.scss:
```diff
@@ -0,0 +1,38 @@
+// It's an EuiFlyout component but shifted to the left side
+// Needs flyout class for specificity
+
+.euiFlyout.euiCollapsibleNav {
```
Is there a better way to do this? Seems pretty rare for us to directly reference another component in sass.

cchaos replied:
We actually have done this is other components. But the only other options are:

1. Lots of `!important`s which also makes it hard for consumers to change
2. Moving the order of the main component index file imports to have collapsible nav come after flyout even though that's not alphabetical.
3. Add another class to the `.euiCollapsibleNav` element just for increasing specificity

snide commented on src/components/collapsible_nav/collapsible_nav.tsx:
```diff
@@ -0,0 +1,75 @@
+  );
+
+  return (
+    <EuiFlyout
```
Likely needs to be wrapped in a portal, since `EuiFlyout` does not do it natively?

cchaos replied:
I'm not sure that's necessary. I'm thinking about where it ends up being used in the header (looking at the prototype) and for tab order (especially when docked) it makes more sense for it to appear within the header not at the bottom of the DOM.

snide replied:
Think you'll run into z-index problems if anything is open in portals underneath it though (like another flyout (which are often wrapped in portals), or a modal).

cchaos replied:
Well I honestly think that's what the consuming application needs to handle. Like a service that monitors all the floating paneling to ensure they don't overlap.

The main problem I see with a portal is the place in the DOM and tab order.

```
body
-> header
-> page content
-> portalled navigation
```

We'd need to manually manage tab order from header to navigation and manage it back from navigation to page content. It's really hard to maintain from a component perspective that has 0 knowledge of the consuming application structure.

myasonik replied:
++

Agreed that dealing with multiple overlapping things sounds like a consuming app problem

snide commented on src-docs/src/views/collapsible_nav/collapsible_nav_example.js:
```diff
@@ -0,0 +1,72 @@
+            <EuiCode>dock</EuiCode> the flyout. This affixes the flyout to the
+            window and pushes the body content by adding left side padding.
+          </p>
+          <EuiCallOut title="Smaller screen sizes don't allow for docking">
```
```suggestion
          <EuiCallOut title="Docking is not possible on small screens because it would force less real estate for the page content">
```

Think this would allow you to remove the below paragraph and leave it a title only.

snide commented on src/components/collapsible_nav/_collapsible_nav.scss:
```diff
@@ -0,0 +1,38 @@
+  animation: euiCollapsibleNavIn $euiAnimSpeedNormal $euiAnimSlightResistance;
+}
+
+@include euiBreakpoint('l', 'xl') {
```
Should the nav become full width on small breakpoints?

cchaos replied:
I don't usually live nav drawers that take over the full screen as it's not clear that I didn't go to another page vs an overlay. But this does bring up the idea that we'll need a "Close nav" button on tiny screens since the overaly mask will only be a tiny bit visible/clickable.

myasonik replied:
Maybe on narrow widths it becomes docked at the top of the page? (So you can keep scrolling down past it)

cchaos replied:
Ok I think we're actually on the same page. I noticed that with my brute force max-width overrides, it doesn't shrink on smaller-than-the-width-of-the-flyout screens. You were pointing out a bug not **asking** for it to actually become full width.

cchaos replied:
> Maybe on narrow widths it becomes docked at the top of the page? (So you can keep scrolling down past it)

Honestly, that's how our EUI docs work, and I find it really frustrating.

cchaos commented on src-docs/src/views/collapsible_nav/collapsible_nav_example.js:
```diff
@@ -0,0 +1,72 @@
+            <strong>EuiCollapsibleNav</strong> is simply a custom wrapper around{' '}
+            <Link to="/layout/flyout">
+              <strong>EuiFlyout</strong>
+            </Link>{' '}
```
Perhaps? I'm terrible at semicolon usage knowledge.

cchaos commented on src-docs/src/views/collapsible_nav/collapsible_nav_example.js:
```diff
@@ -0,0 +1,72 @@
+          <p>
+            <strong>EuiCollapsibleNav</strong> is simply a custom wrapper around{' '}
+            <Link to="/layout/flyout">
+              <strong>EuiFlyout</strong>
```
Yes, I have a personal task to change all references to components to use the strong tag instead of EuiCode. My reasoning is that the name of the component then gets jumbled with the styling of the actual props. It's harder to discern what is a prop and what is the name of the thing. It also makes linking to other components more consistent since it's weird to link a code block.

myasonik commented on src/components/collapsible_nav/collapsible_nav.tsx:
```diff
@@ -0,0 +1,75 @@
+  );
+
+  return (
+    <EuiFlyout
```
I'm afraid that we can't reuse `flyout` wholesale like this... Flyout has dialog semantics and a focus trap built into it. This works well enough for the undocked state but for the docked state we need to drop the dialog semantics (removing `role="dialog"` and `tabIndex="0"`).

cchaos replied:
Perhaps then to satisfy this and @snide's comment about SASS overrides, I'll just make a plain ole div and re-use flyout styling via a new mixin.

myasonik commented on src/components/collapsible_nav/collapsible_nav.tsx:
```diff
@@ -0,0 +1,75 @@
+
+  return (
+    <EuiFlyout
+      ownFocus={!isDocked}
```
This didn't seem to be working for me in the demo site (I was focus trapped no matter if I was docked or not)

cchaos replied:
`ownFocus` just adds the `overlayMask` (white area) where clicking the mask will close the navigation.

cchaos replied:
It doesn't focus-trap

myasonik replied:
Ah, makes sense. 

This issue should just fall away if you plan on moving away from `EuiFlyout` so all's good

myasonik commented on src/components/collapsible_nav/collapsible_nav.tsx:
```diff
@@ -0,0 +1,75 @@
+      className={classes}
+      hideCloseButton={true}
+      {...rest}>
+      {children}
```
Can we wrap this in `<nav />` and require that users pass in either an `aria-label` or `aria-labelledby` for it?

myasonik replied:
Hmm... In a perfect world I'd say no... 

But I wonder if it's better to have that compromise and guarantee that a `nav` is always present rather than the flip side compromise of pushing the responsibility to add a `nav` to implementing devs...

As an escape hatch, there could be a "children" and an "otherChildren" prop... Which has some discoverability and ergonomic issues...

Just thinking out loud and don't really have a great solution here. Any thoughts?

myasonik commented on src/components/collapsible_nav/collapsible_nav.tsx:
```diff
@@ -0,0 +1,75 @@
+      onClose={collapse}
+      size="s"
+      className={classes}
+      hideCloseButton={true}
```
In the prop docs you mention that if you hide the close button the implementing developer must add another one somewhere in the flyout (which is a great rule!) but then forget to add one to the example which makes me thing many implementing developers will do the same.

I wonder about removing the dialog semantics all together (never having `role="dialog"`, `tabIndex="0"` or the focus trap) and instead allowing users to tab through and out of the collapsible nav (and on tab out, the nav closes).

This would remove some semantic complexity and remove a foot-gun from implementing devs.

cchaos replied:
> but then forget to add one to the example

Not sure I follow. The EuiCollapsibleNav example contains a button that opens **and closes** the flyout, plus it uses `ownFocus` which allows clicking the area outside of the flyout to close it.

Controlling visibility via tab focus has been huge pain to get right with the current EuiNavDrawer. If someone can figure out a way to really nail it, I'm down for the idea. (read as: Go for it! I will not be tackling that challenge 😉)

myasonik replied:
👍 I'll take a stab at it after [some of the other changes](https://github.com/elastic/eui/pull/2977#discussion_r387938521) we talked about go in

cchaos commented:
> When undocking the nav, I was surprised when the nav was still open. What do you think about closing the nav when undocked?

The current behavior is how it was working during the Usability tests. We didn't hear any feedback specifically on this interaction so I'm not inclined to change it.

cchaos commented:
I should also mention that the example for CollapsibleNav is **not complete**. It will get fleshed out with all the other inner content, but for now I'm just trying to nail the actual "drawer" part.

thompsongl commented:
> Is there any way to add tests for the responsive behavior? Can it mock setting the window to a certain size?

I'm not sure that any `@media` queries would get used (jsdom does not actually draw anything), but render changes based on `window.innerwidth` in JS _should_ get picked with something like

```js
const resizeWindow = (width, height) => {
  window.innerWidth = width
  window.innerHeight = height
  window.dispatchEvent(new Event('resize'))
}
```

cchaos commented:
Ok the component has been updated to not render a EuiFlyout, but just reuses some styles and copies some functionality over. I addressed (most) of the other comments as well.

I also added a visable "close" button (though I supposed we could make it a accessibleOnFocus element instead).

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2977/

myasonik commented:
Is there any way to have the portal render at the start of the document for when rendering the docked version of the nav? Having it at the end of the document, I think, is pretty confusing from a tab order perspective.

cchaos commented:
> Is there any way to have the portal render at the start of the document for when rendering the docked version of the nav? Having it at the end of the document, I think, is pretty confusing from a tab order perspective.

The component does **not** live in a portal. It's DOM will get inserted wherever you place the component. The **example** in the EUI docs is at the tail end of the DOM structure so that's where it gets rendered.

cchaos commented:
> I'd suggest setting up a PR for the feature branch merge (in draft mode)

Yep, I have to have new commits in the branch before Github allows me to create a PR, so I need this one in first.
